### PR TITLE
chore: better github installation flow

### PR DESCRIPTION
**Pull request by alaister**
- skips the github product page
- skips org selection page once gh integration is installed
- auto selects project in linker
- redirects back to project & opens branching installation modal

### CODE REVIEW

File: studio/components/interfaces/Integrations/ProjectLinker.tsx [hunk 1]
```diff
@@ -79,9 +79,18 @@ const ProjectLinker = ({
   const [supabaseProjectRef, setSupabaseProjectRef] = useState<string | undefined>(
     defaultSupabaseProjectRef
   )
+  useEffect(() => {
+    if (defaultSupabaseProjectRef !== undefined && supabaseProjectRef == undefined)
+      setSupabaseProjectRef(defaultSupabaseProjectRef)
+  }, [defaultSupabaseProjectRef, supabaseProjectRef])
+
```
Comment (MildTomato) [inline studio/components/interfaces/Integrations/ProjectLinker.tsx:86]: Couldn't all this logic just go directly in the useState initialState?

Comment (alaister) [reply studio/components/interfaces/Integrations/ProjectLinker.tsx:86]: That would be nice, but `defaultSupabaseProjectRef` is `undefined `initially thanks to it being a next.js query param

Comment (MildTomato) [reply studio/components/interfaces/Integrations/ProjectLinker.tsx:86]: gotcha

File: studio/pages/project/[ref]/index.tsx [hunk 2]
```diff
@@ -6,11 +9,24 @@ import ProjectPausedState from 'components/layouts/ProjectLayout/ProjectPausedSt
...
 
 const Home: NextPageWithLayout = () => {
   const project = useSelectedProject()
 
+  const snap = useAppStateSnapshot()
+  const { enabledBranching } = useParams()
+
+  const hasShownEnableBranchingModalRef = useRef(false)
+  useEffect(() => {
+    if (enabledBranching && !hasShownEnableBranchingModalRef.current) {
```
Comment (joshenlim) [inline studio/pages/project/[ref]/index.tsx:23]: hmm this block isn't getting triggered for some reason - am getting redirected to the home page of the project ref, but the modal isn't opening

Comment (joshenlim) [reply studio/pages/project/[ref]/index.tsx:23]: separately though - imo a better UX would be to close the tab after settling the github integrations, and update the state of the tab that was already open

cause now i've got 2 tabs of the project open after selecting a GH repo to integrate with

File: studio/components/layouts/AppLayout/EnableBranchingButton/GithubRepositorySelection.tsx [hunk 3]
```diff
@@ -106,7 +108,11 @@ const GithubRepositorySelection = ({
             </Link>
           )}
           {hasGithubIntegrationInstalled && !githubConnection && (
-            <EmptyIntegrationConnection showNode={false} onClick={() => onSelectConnectRepo()} />
+            <EmptyIntegrationConnection
```
Comment (joshenlim) [inline studio/components/layouts/AppLayout/EnableBranchingButton/GithubRepositorySelection.tsx:111]: Not this line - but within this file

"Configure connection" button opens the side panel, but shows "No Vercel projects available" for a second before showing the GH repo picker - Needs an appropriate loading state here

Uploading Screen Recording 2023-10-16 at 16.07.56.mov…

File: studio/components/interfaces/Integrations/ProjectLinker.tsx [hunk 4]
```diff
@@ -79,9 +79,18 @@ const ProjectLinker = ({
   const [supabaseProjectRef, setSupabaseProjectRef] = useState<string | undefined>(
     defaultSupabaseProjectRef
   )
+  useEffect(() => {
+    if (defaultSupabaseProjectRef !== undefined && supabaseProjectRef == undefined)
```
Comment (joshenlim) [inline studio/components/interfaces/Integrations/ProjectLinker.tsx:83]: tiny nit:
```suggestion
    if (defaultSupabaseProjectRef !== undefined && supabaseProjectRef === undefined)
```

File: studio/pages/integrations/github/install.tsx [hunk 5]
```diff
@@ -48,18 +49,27 @@ const GitHubIntegration: NextPageWithLayout = () => {
     },
   })
 
+  const onInstalled = useCallback(
+    (id: string, orgSlug?: string) => {
+      setOrganizationIntegrationId(id)
```
Comment (joshenlim) [inline studio/pages/integrations/github/install.tsx:54]: do we need this? just wondering cause we're changing routes as part of this - the state will get lost either way

File: studio/lib/github-integration.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
+import { useParams } from 'common'
+import { useGitHubIntegrationCreateMutation } from 'data/integrations/github-integration-create-mutation'
+import { useOrganizationsQuery } from 'data/organizations/organizations-query'
+import { useProjectsQuery } from 'data/projects/projects-query'
+import { useEffect, useState } from 'react'
+
+export function useGitHubIntegrationAutoInstall(callback?: (id: string, orgSlug: string) => void) {
+  const { installation_id: installationId, state: projectRef } = useParams()
```
[the remark this answers is not in the document]
Comment (joshenlim) [reply studio/lib/github-integration.ts:8]: i tested locally but i'm not sure if i went down the auto install path

[the remark this answers is not in the document]
Comment (alaister) [reply studio/lib/github-integration.ts:8]: github adds `installation_id` after it redirects to `/integrations/github/install` :)

File: studio/components/layouts/AppLayout/EnableBranchingButton/GithubRepositorySelection.tsx (hunk 3 above)
Comment (alaister) [reply studio/components/layouts/AppLayout/EnableBranchingButton/GithubRepositorySelection.tsx:111]: Nice catch – fixed!

File: studio/pages/project/[ref]/index.tsx (hunk 2 above)
Comment (alaister) [reply studio/pages/project/[ref]/index.tsx]: Fixed!

I've made it open in the same window for now – should reduce the overall number of windows :)

**Review by joshenlim — APPROVED** [review]

**Pull request merged**